File: Vigenère Cipher.py
def generateKey(str, key):
    x = len(str)
    i = 0
    while True:
        if x == i:
            i = 0
        if len(key) == len(str):
            break
        key += key[i]
        i += 1
    return key

def encode(str, key):
    cipherText = ""
    for i in range(len(str)):
        if str[i].isalpha():
            x = (ord(str[i].upper()) + ord(key[i].upper())) % 26
            if str[i].isupper():
                cipherText += chr(x + ord('A'))
            else:
                cipherText += chr(x + ord('a'))
        else:
            cipherText += str[i]
    return cipherText

def decode(cipherText, key):
    origText = ""
    for i in range(len(cipherText)):
        if cipherText[i].isalpha():
            x = (ord(cipherText[i].upper()) - ord(key[i].upper()) + 26) % 26
            if cipherText[i].isupper():
                origText += chr(x + ord('A'))
            else:
                origText += chr(x + ord('a'))
        else:
            origText += cipherText[i]
    return origText

def main():
    while True:
        print("\nVigenère Cipher Program")
        print("1. Encode Text")
        print("2. Decode Text")
        print("3. Exit")
        choice = int(input("Enter Your Choice: "))
        
        if choice == 1 or choice == 2:
            text = input("\nEnter The Text: ")
            key = input("Enter The Key: ")
            
            if len(key) != len(text):
                print("Key Length Must Match The Text Length!")
                continue
            
            keyGenerated = generateKey(text, key)
            
            if choice == 1:
                print("\nEncoded Text:", encode(text, keyGenerated))
            else:
                print("\nDecoded Text:", decode(text, keyGenerated))
        elif choice == 3:
            print("Exiting Program...")
            break
        else:
            print("Invalid Choice! Please Choose Again!")

if __name__ == "__main__":
    main()

File: Vigenère Cipher.cpp
#include <iostream>
#include <string>
using namespace std;

string generateKey(string str, string key) {
  int x = str.size();
  for (int i = 0;; i++) {
    if (x == i)
      i = 0;
    if (key.size() == str.size())
      break;
    key.push_back(key[i]);
  }
  return key;
}

string encode(string str, string key) {
  string cipherText;
  for (int i = 0; i < str.size(); i++) {
    if (isalpha(str[i])) {
      char x = (toupper(str[i]) + toupper(key[i])) % 26;
      x += 'A';
      cipherText.push_back(x);
    } else {
      cipherText.push_back(str[i]);
    }
  }
  return cipherText;
}

string decode(string cipherText, string key) {
  string origText;
  for (int i = 0; i < cipherText.size(); i++) {
    if (isalpha(cipherText[i])) {
      char x = (toupper(cipherText[i]) - toupper(key[i]) + 26) % 26;
      x += 'A';
      origText.push_back(x);
    } else {
      origText.push_back(cipherText[i]);
    }
  }
  return origText;
}

int main() {
  int choice;
  string text, key;

  while (true) {
    cout << "\nVigenère Cipher Program\n";
    cout << "1. Encode Text\n";
    cout << "2. Decode Text\n";
    cout << "3. Exit\n";
    cout << "Enter Your Choice: ";
    cin >> choice;

    if (choice == 1 || choice == 2) {
      cout << "\nEnter The Text: ";
      cin.ignore();
      getline(cin, text);
      cout << "Enter The Key: ";
      cin >> key;

      if (key.length() != text.length()) {
        cout << "Key Length Must Match The Text Length!\n";
        continue;
      }

      string keyGenerated = generateKey(text, key);

      if (choice == 1) {
        cout << "\nEncoded Text: " << encode(text, keyGenerated) << endl;
      } else {
        cout << "\nDecoded Text: " << decode(text, keyGenerated) << endl;
      }
    } else if (choice == 3) {
      cout << "Exiting Program...\n";
      break;
    } else {
      cout << "Invalid Choice! Please Choose Again!\n";
    }
  }
  return 0;
}
